packages/storage-service/README.md

```diff
@@ -0,0 +1,377 @@
+const service = new StorageService({
+  messenger: testMessenger,
+  // storage: undefined, // Optional - defaults to InMemoryStorageAdapter
+});
+
+// Works immediately, data isolated per test
+await service.setItem('TestController', 'key', 'value');
+```
+
+#### 3. Delegate Actions to Controllers
+
+```typescript
+rootMessenger.delegate({
+  actions: [
+    'StorageService:setItem',
+    'StorageService:getItem',
+    'StorageService:removeItem',
+  ],
+  messenger: snapControllerMessenger,
+});
+```
+
+### Direct Usage
+
+You can also use the service directly without a messenger:
+
+```typescript
+import { StorageService, InMemoryStorageAdapter } from '@metamask/storage-service';
+
+const service = new StorageService({
+  messenger: myMessenger,
+  storage: new InMemoryStorageAdapter(),
+});
+
+await service.setItem('MyController', 'myKey', { data: 'value' });
+const data = await service.getItem('MyController', 'myKey');
+```
+
+## API
+
+### `StorageService`
+
+#### `setItem<T>(namespace: string, key: string, value: T): Promise<void>`
+
+Store data in storage.
+
+- `namespace` - Controller namespace (e.g., 'SnapController')
+- `key` - Storage key (e.g., 'npm:@metamask/bitcoin-wallet-snap:sourceCode')
+- `value` - Data to store (will be JSON stringified)
+
+```typescript
+await service.setItem('SnapController', 'snap-id:sourceCode', sourceCode);
+```
+
+#### `getItem<T>(namespace: string, key: string): Promise<T | null>`
+
+Retrieve data from storage.
+
+- `namespace` - Controller namespace
+- `key` - Storage key
+- **Returns**: Parsed data or null if not found
+
+```typescript
+const sourceCode = await service.getItem('SnapController', 'snap-id:sourceCode');
+```
+
+#### `removeItem(namespace: string, key: string): Promise<void>`
+
+Remove data from storage.
+
+```typescript
+await service.removeItem('SnapController', 'snap-id:sourceCode');
+```
+
+#### `getAllKeys(namespace: string): Promise<string[]>`
+
+Get all keys for a namespace (without prefix).
+
+```typescript
+const keys = await service.getAllKeys('SnapController');
+// Returns: ['snap-id-1:sourceCode', 'snap-id-2:sourceCode', ...]
+```
+
+#### `clear(namespace: string): Promise<void>`
+
+Clear all data for a namespace.
+
+```typescript
+await service.clear('SnapController');
+```
+
+## StorageAdapter Interface
+
+Implement this interface to provide platform-specific storage. Adapters are responsible for:
+- Building the full storage key (e.g., `storageService:namespace:key`)
+- Wrapping data with metadata (timestamp) before serialization
+- Serializing/deserializing data (JSON.stringify/parse)
+
+```typescript
+export type StorageAdapter = {
+  getItem(namespace: string, key: string): Promise<unknown>;
+  setItem(namespace: string, key: string, value: unknown): Promise<void>;
+  removeItem(namespace: string, key: string): Promise<void>;
+  getAllKeys(namespace: string): Promise<string[]>;
+  clear(namespace: string): Promise<void>;
+};
+```
+
+## When to Use
```

Comment:
Nit: Personally I'd put this higher up, this seems important

Reply:
Thanks @Gudahtt, I've improved the readme! Let me know what you think so we can resolve the conversation